README.md
# Note
**This plugin needs maintainers -- please comment on this issue https://github.com/logstash-plugins/logstash-output-influxdb/issues/34 if you would like to help!**


# Logstash Plugin

[![Travis Build Status](https://travis-ci.com/logstash-plugins/logstash-output-influxdb.svg)](https://travis-ci.com/logstash-plugins/logstash-output-influxdb)

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Documentation

Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting code or config example, you can use the asciidoc `[source,ruby]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

## Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

## Developing

### 1. Plugin Developement and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-filter-awesome", :path => "/your/local/logstash-filter-awesome"
```
- Install plugin
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Run Logstash with your plugin
```sh
bin/logstash -e 'filter {awesome {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-filter-awesome.gemspec
```
- Install the plugin from the Logstash home
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.


CHANGELOG.md
## 5.0.6
  - [Doc] Fixed typo in `user_event_fields_for_data_points` config option [#95](https://github.com/logstash-plugins/logstash-output-influxdb/pull/95)

## 5.0.5
  - Added clarifying information for `send_as_tags` config option [#80](https://github.com/logstash-plugins/logstash-output-influxdb/pull/80)
  
## 5.0.4
  - Docs: Set the default_codec doc attribute.

## 5.0.3
  - Update gemspec summary

## 5.0.2
  - Fix some documentation issues
  
## 5.0.1
  - Fix a typo in exception handling 
  - Various build and ci changes

## 5.0.0 
 - Use the official influxdb client. This doesn't change the config options, but has some
   breakage potential, so this is a major version bump.

## 4.0.0
 - Breaking: Updated plugin to use new Java Event APIs
 - Fix the suite related to the new java event and the ordering of the keys in the hash.
 - Relax logstash-core-plugin-api constraints
 - update .travis.yml

## 3.1.2
 - Depend on logstash-core-plugin-api instead of logstash-core, removing the need to mass update plugins on major releases of logstash

## 3.1.1
 - New dependency requirements for logstash-core for the 5.0 release

## 3.1.0
 - New option to enable SSL/TLS encrypted communication to InfluxDB
 - DB parameter now supports sprintf formatting

## 3.0.0
 - Update to work with influxdb 0.9
 - breaking change, renaming 'series' to 'measurement'
 - breaking change for values of time_precision
 - Special characters now properly escaped in tag key/value, field key, measurement

## 2.0.0
 - Plugins were updated to follow the new shutdown semantic, this mainly allows Logstash to instruct input plugins to terminate gracefully, 
   instead of using Thread.raise on the plugins' threads. Ref: https://github.com/elastic/logstash/pull/3895
 - Dependency on logstash-core update to 2.0



spec/outputs/influxdb_spec.rb
# encoding: utf-8
require "logstash/devutils/rspec/spec_helper"
require "logstash/outputs/influxdb"

describe LogStash::Outputs::InfluxDB do

  subject { LogStash::Outputs::InfluxDB.new(config) }
    
  context "validate minimal default config" do

    let(:config) do
    {
      "host" => "testhost",
      "use_event_fields_for_data_points" => true
    }
    end

    before do
      subject.register
      subject.close
    end

    it "sets correct influx client settings" do

      config = subject.instance_variable_get(:@influxdbClient).config

      expect(config.next_host).to eq "testhost"  
      expect(config.instance_variable_get(:@port)).to eq 8086 
      expect(config.instance_variable_get(:@time_precision)).to eq "ms" 
      expect(config.instance_variable_get(:@auth_method)).to eq "none".freeze 
      expect(config.instance_variable_get(:@initial_delay)).to eq 1
      expect(config.instance_variable_get(:@retry)).to eq 3
      expect(config.instance_variable_get(:@use_ssl)).to eq false
      expect(config.instance_variable_get(:@username)).to eq nil 
      expect(config.instance_variable_get(:@password)).to eq nil 

    end

  end

  context "validate non default config" do

    let(:config) do
    {
      "host" => "localhost",
      "use_event_fields_for_data_points" => true,
      "port" => 9999,
      "ssl" => true,
      "user" => "my_user",
      "password" => "my_pass",
      "initial_delay" => 5,
      "max_retries" => 8,
      "time_precision" => "s"
    }
    end

    before do
      subject.register
      subject.close
    end

    it "sets correct influx client settings" do
      config = subject.instance_variable_get(:@influxdbClient).config
      expect(config.instance_variable_get(:@port)).to eq 9999
      expect(config.instance_variable_get(:@time_precision)).to eq "s"   
      expect(config.instance_variable_get(:@initial_delay)).to eq 5
      expect(config.instance_variable_get(:@retry)).to eq 8
      expect(config.instance_variable_get(:@use_ssl)).to eq true    
      expect(config.instance_variable_get(:@username)).to eq "my_user"
      expect(config.instance_variable_get(:@password)).to eq "my_pass"  
      expect(config.instance_variable_get(:@auth_method)).to eq "params".freeze           
    end

  end
  
  context "complete pipeline run with 2 events" do

    let(:config) do
    {
      "host" => "localhost",
      "user" => "someuser",
      "password" => "somepwd",
      "allow_time_override" => true,
      "data_points" => {
        "foo" => "%{foo}",
        "bar" => "%{bar}",
        "time" => "%{time}"
      }
    }
    end

    before do
      subject.register
      # Added db name parameter to post - M.Laws
      allow(subject).to receive(:dowrite).with(result, "statistics")

      2.times do
        subject.receive(LogStash::Event.new("foo" => "1", "bar" => "2", "time" => "3", "type" => "generator"))
      end

      # Close / flush the buffer
      subject.close
    end

    #let(:result) { "logstash foo=\"1\",bar=\"2\" 3\nlogstash foo=\"1\",bar=\"2\" 3" }

    let(:result) {[{:series=>"logstash", :timestamp=>"3", :values=>{"foo"=>"1", "bar"=>"2"}}, {:series=>"logstash", :timestamp=>"3", :values=>{"foo"=>"1", "bar"=>"2"}}]}

    it "should receive 2 events, flush and call post with 2 items json array" do
      expect(subject).to have_received(:dowrite).with(result, "statistics")
    end

  end

  context "using event fields as data points" do

    let(:config) do
    {
      "host" => "localhost",
      "measurement" => "my_series",
      "allow_time_override" => true,
      "use_event_fields_for_data_points" => true
    }
    end

    before do
      subject.register
      # Added db name parameter to post - M.Laws
      allow(subject).to receive(:dowrite).with(result, "statistics")

      subject.receive(LogStash::Event.new("foo" => "1", "bar" => "2", "time" => "3", "type" => "generator"))

      # Close / flush the buffer
      subject.close
    end

    let(:result) {[{:series=>"my_series", :timestamp=>"3", :values=>{"foo"=>"1", "bar"=>"2"}}]}

    it "should use the event fields as the data points, excluding @version and @timestamp by default as well as any fields configured by exclude_fields" do
      expect(subject).to have_received(:dowrite).with(result, "statistics")
    end

  end


  context "sending some fields as Influxdb tags" do

    let(:config) do
    {
      "host" => "localhost",
      "measurement" => "my_series",
      "allow_time_override" => true,
      "use_event_fields_for_data_points" => true,
      "send_as_tags" => ["bar", "baz", "qux"]
    }
    end

    before do
      subject.register
      # Added db name parameter to post - M.Laws
      allow(subject).to receive(:dowrite).with(result, "statistics")

      subject.receive(LogStash::Event.new("foo" => "1", "bar" => "2", "baz" => "3", "time" => "4", "type" => "generator"))

      # Close / flush the buffer
      subject.close
    end

    let(:result) {[{:series=>"my_series", :timestamp=>"4", :tags=>{"bar"=>"2", "baz"=>"3"}, :values=>{"foo"=>"1"}}]}

    it "should use the event fields as the data points, excluding @version and @timestamp by default as well as any fields configured by exclude_fields" do
      expect(subject).to have_received(:dowrite).with(result, "statistics")
    end

  end

  context "when fields data contains a list of tags" do

    let(:config) do
    {
      "host" => "localhost",
      "measurement" => "my_series",
      "allow_time_override" => true,
      "use_event_fields_for_data_points" => true,
    }
    end

    before do
      subject.register
      # Added db name parameter to post - M.Laws
      allow(subject).to receive(:dowrite).with(result, "statistics")

      subject.receive(event)

      # Close / flush the buffer
      subject.close
    end
    
    let(:event) {LogStash::Event.new("foo" => "1", "time" => "2", "tags" => ["tagged"], "type" => "generator")}
    let(:result) {[{:series=>"my_series", :timestamp=>"2", :tags=>{"tagged"=>"true"}, :values=>{"foo"=>"1"}}]}

    it "should use the event fields as the data points, excluding @version and @timestamp by default as well as any fields configured by exclude_fields" do
      expect(subject).to have_received(:dowrite).with(result, "statistics")
    end

  end

  context "read data-points/tags from speicifc event fields" do

    let(:config) do
    {
      "host" => "localhost",
      "measurement" => "my_series",
      "allow_time_override" => true,
      "use_event_fields_for_data_points" => true,
      "use_hash_field_for_data_points" => "[pts]",
      "send_array_field_as_tags" => "[tags]",
    }
    end

    before do
      subject.register
      allow(subject).to receive(:dowrite).with(result, "statistics")
      subject.receive(event)
      subject.close
    end

    let(:event) {LogStash::Event.new("pts" => {"foo" => "bar","fee" =>"buzz","time" => 22}, "tags" => ["foo"], "type" => "generator")}
    let(:result) {[{:series=>"my_series", :timestamp=>22, :tags=>{"foo"=>"bar"}, :values=>{"fee"=>"buzz"}}]}

    it "should use the hash field entries as the data points and array field as tags" do
      expect(subject).to have_received(:dowrite).with(result, "statistics")
    end

  end

  context "when fields are coerced to numerics" do

    let(:config) do
    {
      "host" => "localhost",
      "measurement" => "my_series",
      "allow_time_override" => true,
      "use_event_fields_for_data_points" => true,
      "coerce_values" => { "foo" => "integer", "bar" => "float" }
    }
    end

    before do
      subject.register
      # Added db name parameter to post - M.Laws
      allow(subject).to receive(:dowrite).with(result, "statistics")

      subject.receive(LogStash::Event.new("foo" => "1", "bar" => "2.0", "baz"=>"\\\"quotes\\\"", "time"=>3, "type" => "generator"))

      # Close / flush the buffer
    subject.close
    end

    let(:result) {[{:series=>"my_series", :timestamp=>3, :values=>{"foo"=>1, "bar"=>2.0, "baz"=>"\\\"quotes\\\"" }}]}

    it "should use the event fields as the data points, excluding @version and @timestamp by default as well as any fields configured by exclude_fields" do
      expect(subject).to have_received(:dowrite).with(result, "statistics")
    end

  end

  # Test issue #31 - Run "db" parameter through event.sprintf() to support...
  # -------------------------------------------------------------------------
  # This test is intended to verify writes to multiple measurements in A SINGLE
  # DATABASE continue to work *after* implementing #31.  Also verifies that
  # sprintf formatting is supported in the measurement name.
  context "receiving 3 points between 2 measurements in 1 database" do

    let(:config) do
    {
      "host" => "localhost",
      "measurement" => "%{baz}",
      "allow_time_override" => true,
      "use_event_fields_for_data_points" => true
    }
    end

    before do
      subject.register
      # Added db name parameter to post - M.Laws
      allow(subject).to receive(:dowrite)

      subject.receive(LogStash::Event.new("foo"=>"1", "bar"=>"2", "baz" => "m1", "time" => "1", "type" => "generator"))
      subject.receive(LogStash::Event.new("foo"=>"3", "bar"=>"4", "baz" => "m2", "time" => "2", "type" => "generator"))
      subject.receive(LogStash::Event.new("foo"=>"5", "bar"=>"6", "baz" => "m2", "time" => "3", "type" => "generator"))

      # Close / flush the buffer
      subject.close
    end

    let(:result) {[{:series=>"m1", :timestamp=>"1", :values=>{"foo"=>"1", "bar"=>"2", "baz" => "m1" }},{:series=>"m2", :timestamp=>"2", :values=>{"foo"=>"3", "bar"=>"4", "baz" => "m2"}},{:series=>"m2", :timestamp=>"3", :values=>{"foo"=>"5", "bar"=>"6", "baz" => "m2" }}]}

    it "should use the event fields as the data points, excluding @version and @timestamp by default as well as any fields configured by exclude_fields" do
      expect(subject).to have_received(:dowrite).with(result, "statistics")
    end

  end

  # Test issue #31 - Run "db" parameter through event.sprintf() to support...
  # -------------------------------------------------------------------------
  # This test is intended to verify writes to multiple measurements in MULTIPLE
  # DATABASES result in separate bulk POSTs (one for each database in the
  # buffer), and the correct measurements being written to the correct db.
  # Also verifies that sprintf formatting is correctly supported in the
  # database name.
  context "receiving 4 points between 2 measurements in 2 databases" do

    let(:config) do
    {
      "host" => "localhost",
      "db" => "%{bar}",
      "measurement" => "%{baz}",
      "allow_time_override" => true,
      "use_event_fields_for_data_points" => true,
    }
    end

    before do
      subject.register
      # Added db name parameter to post - M.Laws
      allow(subject).to receive(:dowrite)

      subject.receive(LogStash::Event.new("foo"=>"1", "bar"=>"db1", "baz" => "m1", "time" => "1", "type" => "generator"))
      subject.receive(LogStash::Event.new("foo"=>"2", "bar"=>"db1", "baz" => "m1", "time" => "2", "type" => "generator"))
      subject.receive(LogStash::Event.new("foo"=>"3", "bar"=>"db2", "baz" => "m2", "time" => "3", "type" => "generator"))
      subject.receive(LogStash::Event.new("foo"=>"4", "bar"=>"db2", "baz" => "m2", "time" => "4", "type" => "generator"))
      # Close / flush the buffer
      subject.close
    end

    let(:resultdb1) {[{:series=>"m1", :timestamp=>"1", :values=>{"foo"=>"1", "bar"=>"db1", "baz" => "m1" }},{:series=>"m1", :timestamp=>"2", :values=>{"foo"=>"2", "bar"=>"db1", "baz" => "m1" }}]}
    let(:resultdb2) {[{:series=>"m2", :timestamp=>"3", :values=>{"foo"=>"3", "bar"=>"db2", "baz" => "m2" }},{:series=>"m2", :timestamp=>"4", :values=>{"foo"=>"4", "bar"=>"db2", "baz" => "m2" }}]}

    it "should use the event fields as the data points, excluding @version and @timestamp by default as well as any fields configured by exclude_fields" do
      expect(subject).to have_received(:dowrite).with(resultdb1, "db1").once
      expect(subject).to have_received(:dowrite).with(resultdb2, "db2").once
    end


  end
end


lib/logstash/outputs/influxdb.rb
# encoding: utf-8
require "logstash/namespace"
require "logstash/outputs/base"
require "logstash/json"
require "stud/buffer"
require "influxdb"

# This output lets you output Metrics to InfluxDB (>= 0.9.0-rc31)
#
# The configuration here attempts to be as friendly as possible
# and minimize the need for multiple definitions to write to
# multiple measurements and still be efficient
#
# the InfluxDB API let's you do some semblance of bulk operation
# per http call but each call is database-specific
#
# You can learn more at http://influxdb.com[InfluxDB homepage]
class LogStash::Outputs::InfluxDB < LogStash::Outputs::Base
  include Stud::Buffer

  config_name "influxdb"

  # The database to write - supports sprintf formatting
  config :db, :validate => :string, :default => "statistics"

  # The retention policy to use
  config :retention_policy, :validate => :string, :default => "autogen"

  # The hostname or IP address to reach your InfluxDB instance
  config :host, :validate => :string, :required => true

  # The port for InfluxDB
  config :port, :validate => :number, :default => 8086

  # The user who has access to the named database
  config :user, :validate => :string, :default => nil

  # The password for the user who access to the named database
  config :password, :validate => :password, :default => nil

  # Enable SSL/TLS secured communication to InfluxDB
  config :ssl, :validate => :boolean, :default => false

  # Measurement name - supports sprintf formatting
  config :measurement, :validate => :string, :default => "logstash"

  # Hash of key/value pairs representing data points to send to the named database
  # Example: `{'column1' => 'value1', 'column2' => 'value2'}`
  #
  # Events for the same measurement will be batched together where possible
  # Both keys and values support sprintf formatting
  config :data_points, :validate => :hash, :default => {}, :required => true

  # Allow the override of the `time` column in the event?
  #
  # By default any column with a name of `time` will be ignored and the time will
  # be determined by the value of `@timestamp`.
  #
  # Setting this to `true` allows you to explicitly set the `time` column yourself
  #
  # Note: **`time` must be an epoch value in either seconds, milliseconds or microseconds**
  config :allow_time_override, :validate => :boolean, :default => false

  # Set the level of precision of `time`
  #
  # only useful when overriding the time value
  config :time_precision, :validate => ["n", "u", "ms", "s", "m", "h"], :default => "ms"

  # Allow value coercion
  #
  # this will attempt to convert data point values to the appropriate type before posting
  # otherwise sprintf-filtered numeric values could get sent as strings
  # format is `{'column_name' => 'datatype'}`
  #
  # currently supported datatypes are `integer` and `float`
  #
  # supports sprintf-formatting in column names
  #
  config :coerce_values, :validate => :hash, :default => {}

  # Automatically use fields from the event as the data points sent to Influxdb
  config :use_event_fields_for_data_points, :validate => :boolean, :default => false
  
  # An array containing the names of fields from the event to exclude from the
  # data points 
  # 
  # Events, in general, contain keys "@version" and "@timestamp". Other plugins
  # may add others that you'll want to exclude (such as "command" from the 
  # exec plugin).
  # 
  # This only applies when use_event_fields_for_data_points is true.
  config :exclude_fields, :validate => :array, :default => ["@timestamp", "@version", "sequence", "message", "type"]  

  # An array containing the names of fields to send to Influxdb as tags instead
  # of values. Influxdb 0.9 convention is that values that do not change every
  # request should be considered metadata and given as tags.
  # Tags are only sent when present in `data_points` or if `use_event_fields_for_data_points` is `true`.
  config :send_as_tags, :validate => :array, :default => ["host"]

  # This setting controls how many events will be buffered before sending a batch
  # of events. Note that these are only batched for the same measurement
  config :flush_size, :validate => :number, :default => 100

  # The amount of time since last flush before a flush is forced.
  #
  # This setting helps ensure slow event rates don't get stuck in Logstash.
  # For example, if your `flush_size` is 100, and you have received 10 events,
  # and it has been more than `idle_flush_time` seconds since the last flush,
  # logstash will flush those 10 events automatically.
  #
  # This helps keep both fast and slow log streams moving along in
  # near-real-time.
  config :idle_flush_time, :validate => :number, :default => 1
  
  # The amount of time in seconds to delay the initial retry on connection failure.  
  #
  # The delay will increase exponentially for each retry attempt (up to max_retries).
  
  config :initial_delay, :validate => :number, :default => 1
  
  # The number of time to retry recoverable errors before dropping the events.
  #
  # A value of -1 will cause the plugin to retry indefinately.
  # A value of 0 will cause the plugin to never retry.
  # Otherwise it will retry up to the specified mumber of times.
  #
  config :max_retries, :validate => :number, :default => 3

  public
  def register
    require 'cgi'
    
    @queue = []

    buffer_initialize(
      :max_items => @flush_size,
      :max_interval => @idle_flush_time,
      :logger => @logger
    )
    @auth_method = @user.nil? ? 'none'.freeze  : "params".freeze 
    
    @influxdbClient = InfluxDB::Client.new host: @host, port: @port, time_precision: @time_precision, use_ssl: @ssl, verify_ssl: false, retry: @max_retries, initial_delay: @initial_delay, auth_method: @auth_method, username: @user, password: @password.value
  end # def register


  public
  def receive(event)
    @logger.debug? and @logger.debug("Influxdb output: Received event: #{event}")

    # An Influxdb 0.9 event looks like this: 
    # cpu_load_short,host=server01,region=us-west value=0.64 1434055562000000000
    #  ^ measurement  ^ tags (optional)            ^ fields   ^ timestamp (optional)
    # 
    # Since we'll be buffering them to send as a batch, we'll only collect
    # the values going into the points array
    
    time  = timestamp_at_precision(event.timestamp, @time_precision.to_sym)
    point = create_point_from_event(event)
    exclude_fields!(point)
    coerce_values!(point, event)

    if point.has_key?('time')
      unless @allow_time_override
        @logger.error("Cannot override value of time without 'allow_time_override'. Using event timestamp")
      else
        time = point.delete("time")
      end
    end

    if point.empty?
      @logger.debug? and @logger.debug("Received data points empty")
      return
    end

    tags, point = extract_tags(point)

    event_hash = {
      :series => event.sprintf(@measurement),
      :timestamp       => time,
      :values      => point
    }
    event_hash[:tags] = tags unless tags.empty?

    buffer_receive(event_hash, event.sprintf(@db))
  end # def receive


  def flush(events, database, teardown = false)
    @logger.debug? and @logger.debug("Flushing #{events.size} events to #{database} - Teardown? #{teardown}")
    dowrite(events, database)
  end # def flush
    
  def dowrite(events, database)
    begin
        @influxdbClient.write_points(events, @time_precision, @retention_policy, @db  )
    rescue InfluxDB::AuthenticationError => ae
        @logger.warn("Authentication Error while writing to InfluxDB", :exception => ae)
    rescue InfluxDB::ConnectionError => ce 
        @logger.warn("Connection Error while writing to InfluxDB", :exception => ce)
    rescue Exception => e
        @logger.warn("Non recoverable exception while writing to InfluxDB", :exception => e)
    end
  end

  def close
    buffer_flush(:final => true)
  end # def teardown

 
  # Create a data point from an event. If @use_event_fields_for_data_points is
  # true, convert the event to a hash. Otherwise, use @data_points. Each key and 
  # value will be run through event#sprintf with the exception of a non-String
  # value (which will be passed through)
  def create_point_from_event(event)
    Hash[ (@use_event_fields_for_data_points ? event.to_hash : @data_points).map do |k,v| 
      [event.sprintf(k), (String === v ? event.sprintf(v) : v)] 
    end ]
  end
  

  # Coerce values in the event data to their appropriate type. This requires 
  # foreknowledge of what's in the data point, which is less than ideal. An 
  # alternative is to use a `code` filter and manipulate the individual point's
  # data before sending to the output pipeline
  def coerce_values!(event_data, event)
    @coerce_values.each do |column, value_type|
      column = event.sprintf(column)
      if event_data.has_key?(column)
        begin
          @logger.debug? and @logger.debug("Converting column #{column} to type #{value_type}: Current value: #{event_data[column]}")
          coerced_value, ok = coerce_value(value_type, event_data[column])
          if ok
            event_data[column] = coerced_value
          else
            event_data.delete(column)
          end

        rescue => e
          @logger.error("Unhandled exception", :error => e.message)
        end
      end
    end

    event_data
  end


  def coerce_value(value_type, value)
    case value_type.to_sym
    when :integer
      number = Integer(value) rescue nil
      if number.nil?
        [value, false]
      else
        [number, true]
      end
      
    when :float
      number = Float(value) rescue nil
      if number.nil?
        [value, false]
      else
        [number, true]
      end

    when :string
      [value.to_s, true]
    
    else
      @logger.warn("Don't know how to convert to #{value_type}. Returning value unchanged")
      [value, true]
    end
  end


  # Remove a set of fields from the event data before sending it to Influxdb. This
  # is useful for removing @timestamp, @version, etc
  def exclude_fields!(event_data)
    @exclude_fields.each { |field| event_data.delete(field) }
  end


  # Extract tags from a hash of fields. 
  # Returns a tuple containing a hash of tags (as configured by send_as_tags) 
  # and a hash of fields that exclude the tags. If fields contains a key 
  # "tags" with an array, they will be moved to the tags hash (and each will be
  # given a value of true)
  # 
  # Example: 
  #   # Given send_as_tags: ["bar"]
  #   original_fields = {"foo" => 1, "bar" => 2, "tags" => ["tag"]}
  #   tags, fields = extract_tags(original_fields)
  #   # tags: {"bar" => 2, "tag" => "true"} and fields: {"foo" => 1}
  def extract_tags(fields)
    remainder = fields.dup

    tags = if remainder.has_key?("tags") && remainder["tags"].respond_to?(:inject)
      remainder.delete("tags").inject({}) { |tags, tag| tags[tag] = "true"; tags }
    else
      {}
    end
    
    @send_as_tags.each { |key| (tags[key] = remainder.delete(key)) if remainder.has_key?(key) }

    tags.delete_if { |key,value| value.nil? || value == "" }
    remainder.delete_if { |key,value| value.nil? || value == "" }

    [tags, remainder]
  end


  # Returns the numeric value of the given timestamp in the requested precision.
  # precision must be one of the valid values for time_precision
  def timestamp_at_precision( timestamp, precision )
    multiplier = case precision
      when :h  then 1.0/3600
      when :m  then 1.0/60
      when :s  then 1
      when :ms then 1000
      when :u  then 1000000
      when :n  then 1000000000
    end
    
    (timestamp.to_f * multiplier).to_i
  end

end # class LogStash::Outputs::InfluxDB
